### PR TITLE
Feature/bushfire

### PR DESCRIPTION
**Pull request by samluethi**
full update of the bushfire module (such a complete update that it's now even called wildfire =))

### CODE REVIEW

File: climada/hazard/__init__.py [hunk 1]
```diff
@@ -26,3 +26,5 @@
 from .tc_tracks_forecast import *
 from .tc_rainfield import *
 from .storm_europe import *
+from .wildfire import *
+from .tc_surge import *
```
Comment (emanuel-schmid) [inline climada/hazard/__init__.py:30]:
```suggestion
```

File: climada/entity/impact_funcs/__init__.py [hunk 2]
```diff
@@ -22,4 +22,6 @@
 from .impact_func_set import *
 from .trop_cyclone import *
 from .drought import *
+from .crop_potential import *
```
Comment (emanuel-schmid) [inline climada/entity/impact_funcs/__init__.py:25]:
```suggestion
```

File: climada/entity/impact_funcs/wildfire.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+class IFBushfire(ImpactFunc):
+    """Impact function for bushfire."""
+
+    def __init__(self):
+        ImpactFunc.__init__(self)
+        self.haz_type = 'WF'
+        
+    def set_step(self, threshold):
+        self.haz_type = "WF"
```
Comment (chahank) [inline climada/entity/impact_funcs/wildfire.py:33]: Please add the docstring to these functions. Why the values 295 and 367?

File: climada/entity/impact_funcs/wildfire.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        self.id = 1
+        self.name = "wildfire default"
+        self.intensity_unit = "K"
+        self.intensity = np.array([295, threshold, threshold, 367])
+        self.mdd = np.array([0, 0, 1, 1])
+        self.paa = np.array([1, 1, 1, 1])
+        
+    def set_sigmoid(self, int_range=np.arange(295,500,5), sig_mid=320, sig_shape=0.1, sig_max=1.0):
+        self.haz_type = "WF"
```
Comment (chahank) [inline climada/entity/impact_funcs/wildfire.py:42]: Docstring missing.

File: climada/entity/impact_funcs/wildfire.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        self.id = 1
+        self.name = "wildfire default"
+        self.intensity_unit = "K"
+        self.intensity = int_range
+        self.mdd = sig_max/(1+np.exp(-sig_shape*(int_range-sig_mid)))
+        self.paa = np.ones(len(int_range))
+        
+    def set_default(self, i_half=523.8):
+        self.haz_type = "WF"
```
Comment (chahank) [inline climada/entity/impact_funcs/wildfire.py:51]: Docstring missing. Why 295 - 500? How does this relate to 295 - 367?

File: climada/entity/impact_funcs/wildfire.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+LOGGER = logging.getLogger(__name__)
+
+class IFBushfire(ImpactFunc):
+    """Impact function for bushfire."""
+
+    def __init__(self):
+        ImpactFunc.__init__(self)
+        self.haz_type = 'WF'
+        
+    def set_step(self, threshold):
```
Comment (chahank) [inline climada/entity/impact_funcs/wildfire.py:32]: For all the following: Any reference for these impact functions? How are they calibrated?

File: climada/hazard/wildfire.py [hunk 7]
```diff
@@ -0,0 +1,1158 @@
...
+            centroids (Centroids, optional): centroids in degrees to map data
+        """
+        self.clear()
+
+        # read and initialize data
+        firms = self._clean_firms_csv(csv_firms)
+        # compute centroids
+        res_data = self._firms_resolution(firms)
+        if not centroids:
+            centroids = self._centroids_creation(firms, res_data, centr_res_factor)
```
Comment (chahank) [inline climada/hazard/wildfire.py:134]: Maybe `_centroids_from_firms` ?

File: climada/hazard/wildfire.py [hunk 8]
```diff
@@ -0,0 +1,1158 @@
...
+    def hull_burned_area(self, ev_id, alpha=100.87, return_plot=False):
+        """Compute the burned area for a given fire.
+        Please note: NASA advises against calculating burned area using the
+        FIRMS data.
+
+        Algorithm used: https://pypi.org/project/alphashape/
+
+        Parameters:
+            ev_id: id of the selected fire
+            alpha (float, optional): parameter used to compute the concave hull
```
Comment (chahank) [inline climada/hazard/wildfire.py:173]: I think this should more precise. What does alpha do?

File: climada/hazard/wildfire.py [hunk 9]
```diff
@@ -0,0 +1,1158 @@
...
+        # compute centroids
+        res_data = self._firms_resolution(firms)
+        if not centroids:
+            centroids = self._centroids_creation(firms, res_data, centr_res_factor)
+        else:
+            if not centroids.coord.size:
+                centroids.set_meta_to_lat_lon()
+
+        # define hemisphere
+        if hemisphere is None:
```
Comment (chahank) [inline climada/hazard/wildfire.py:257]: What happens when data is on both sides of the hemisphere?

File: climada/hazard/wildfire.py [hunk 10]
```diff
@@ -0,0 +1,1158 @@
...
+        self.event_name = list(map(str, years))
+        self.date = np.zeros(len(years), int)
+        self.date_end = np.zeros(len(years), int)
+        if hemisphere == 'NHS':
+            for y_idx, year in enumerate(years):
+                self.date[y_idx] = date.toordinal(date(year, 1, 1))
+                self.date_end[y_idx] = date.toordinal(date(year, 12, 31))
+        elif hemisphere == 'SHS':
+            for y_idx, year in enumerate(years):
+                self.date[y_idx] = date.toordinal(date(year, 7, 1))
```
Comment (chahank) [inline climada/hazard/wildfire.py:308]: Probably good to add to the docstring when seasons are in NHS and SHS.

File: climada/hazard/wildfire.py [hunk 11]
```diff
@@ -0,0 +1,1158 @@
...
+        summarized into a probabilistic fire season by calculating the max
+        intensity at each centroid for each probabilistic fire season.
+        Probabilistic fires are created using the logic described in the
+        method 'run_one_bushfire'.
+
+        The fire propagation matrix can be assigned separately, if that is not
+        done it will be generated on the available historic fire (seasons).
+
+        Intensities are drawn randomly from historic events. Thus, this method
+        requires at least one fire to draw from.
```
Comment (chahank) [inline climada/hazard/wildfire.py:334]: This requires a bit more explanation. Either in a future publication, or better already here in the docstring.

File: climada/hazard/wildfire.py [hunk 12]
```diff
@@ -0,0 +1,1158 @@
...
+            self.intensity = sparse.csr_matrix(np.amax(self.intensity, 0))
+            self.event_id = np.array([np.max(self.event_id)+1])
+            self.orig = np.zeros(1, bool)
+            self._set_frequency()
+            self.fraction = self.intensity.copy()
+            self.fraction.data.fill(1.0)
+
+    def summarize_fires_to_seasons(self, year_start=None, year_end=None,
+                                   hemisphere=None):
+        """ Summarize historic fires into fire seasons.
```
Comment (chahank) [inline climada/hazard/wildfire.py:463]: As in previous docstrings, could you add the method how you combine the fires? I.e. by taking the maximum intensity at each centroid.

File: climada/hazard/wildfire.py [hunk 13]
```diff
@@ -0,0 +1,1158 @@
...
+        n_fires = np.zeros(len(years), int)
+        intensity_new = sparse.lil_matrix(np.zeros((len(years), len(self.centroids.lat))))
+
+        for i, year in enumerate(years):
+            if hemisphere == 'NHS':
+                StartTime = date.toordinal(date(year, 1, 1))
+                EndTime = date.toordinal(date(year+1, 1, 1))
+            elif hemisphere == 'SHS':
+                StartTime = date.toordinal(date(year, 7, 1))
+                EndTime = date.toordinal(date(year+1, 7, 1))
```
Comment (chahank) [inline climada/hazard/wildfire.py:500]: This fire season definition looks different from the one in `set_hist_fire_seasons_FIRMS`.

File: climada/hazard/wildfire.py [hunk 14]
```diff
@@ -0,0 +1,1158 @@
...
+            if firms['instrument'].any() == 'MODIS':
+                res_data = 1.0
+            else:
+                res_data = 0.375 # For VIIRS data
+        else:
+            res_data = RES_DATA # For undefined data origin, defined by user
+        return res_data/ONE_LAT_KM
+
+    @staticmethod
+    def _centroids_creation(firms, res_data, centr_res_factor):
```
Comment (chahank) [inline climada/hazard/wildfire.py:597]: Rename to make clear it is only for firms data.

File: climada/hazard/wildfire.py [hunk 15]
```diff
@@ -0,0 +1,1158 @@
...
+        for i in np.unique(firms.event_id):
+            firms.event_id[firms.event_id == i] = event_id_new
+            event_id_new = event_id_new + 1
+
+        firms = firms.reset_index()
+
+        return firms
+
+    @staticmethod
+    def _centroids_resolution(centroids):
```
Comment (chahank) [inline climada/hazard/wildfire.py:755]: This should either be in the `Centroids` module or in the `Util` module.

File: doc/tutorial/climada_hazard_Bushfire-Copy1.ipynb [hunk 16]
```diff
@@ -0,0 +1,452 @@
+{
```
Comment (chahank) [inline doc/tutorial/climada_hazard_Bushfire-Copy1.ipynb:1]: I suppose this document is not needed?

**Review by chahank — COMMENTED** [review]
Just a few quick comments, mostly about docstrings.

**Comment by chahank** [comment]
I think also that the bushfire tutorial should be updated a bit more. Change the name to wildfire. Also, could you add a summary of the data structure at the beginning of the tutorial?

**Comment by emanuel-schmid** [comment]
There is something broken in this branch. Please check the results in [Jenkins ](https://ied-wcr-jenkins.ethz.ch/job/climada_branches/view/change-requests/job/PR-88/).\
In wildfire.py line 48 a module `climada.util.alpha_shape` is expected that does not exist.\
That was removed from the develop branch because it was not used then.\
Either you have to restore it or (preferably) resolve the dependency by a module that is already there.

File: climada/entity/impact_funcs/wildfire.py [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+from climada.entity.impact_funcs.base import ImpactFunc
+
+LOGGER = logging.getLogger(__name__)
+
+class IFBushfire(ImpactFunc):
+    """Impact function for bushfire."""
+
+    def __init__(self):
+        ImpactFunc.__init__(self)
+        self.haz_type = 'WF'
```
Comment (emanuel-schmid) [inline climada/entity/impact_funcs/wildfire.py:30]:
```suggestion
        self.haz_type = 'WF'
        self.id = 1
        self.intensity_unit = "K"
```

File: climada/entity/impact_funcs/wildfire.py [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+        
+    def set_step(self, threshold):
+        self.haz_type = "WF"
+        self.id = 1
+        self.name = "wildfire default"
+        self.intensity_unit = "K"
```
Comment (emanuel-schmid) [inline climada/entity/impact_funcs/wildfire.py:36]:
```suggestion
        self.name = "wildfire step"
```

File: climada/entity/impact_funcs/wildfire.py [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+        self.intensity = np.array([295, threshold, threshold, 367])
+        self.mdd = np.array([0, 0, 1, 1])
+        self.paa = np.array([1, 1, 1, 1])
+        
+    def set_sigmoid(self, int_range=np.arange(295,500,5), sig_mid=320, sig_shape=0.1, sig_max=1.0):
+        self.haz_type = "WF"
+        self.id = 1
+        self.name = "wildfire default"
+        self.intensity_unit = "K"
```
Comment (emanuel-schmid) [inline climada/entity/impact_funcs/wildfire.py:45]:
```suggestion
        self.name = "wildfire sigmoid"
```

File: climada/entity/impact_funcs/wildfire.py [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+        self.intensity = int_range
+        self.mdd = sig_max/(1+np.exp(-sig_shape*(int_range-sig_mid)))
+        self.paa = np.ones(len(int_range))
+        
+    def set_default(self, i_half=523.8):
+        self.haz_type = "WF"
+        self.id = 1
+        self.name = "wildfire default"
+        self.intensity_unit = "K"
```
Comment (emanuel-schmid) [inline climada/entity/impact_funcs/wildfire.py:54]:
```suggestion
        self.name = "wildfire default"
```

File: climada/entity/impact_funcs/wildfire.py (hunk 17 above)
Comment (emanuel-schmid) [reply climada/entity/impact_funcs/wildfire.py:30]: Where is the `id` used? Why is it always `1`? Is it possible to rename it to something better searchable, like `wildfire_impf_id`?

**Review by emanuel-schmid — CHANGES_REQUESTED** [review]
It seems to me that "bushfire" was renamed to "wildfire". Correct?\
If so, please rename everything accordingly, in particular test modules and classes.

File: climada/entity/impact_funcs/test/test_bf.py [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+import numpy as np
+
+from climada.entity.impact_funcs.bushfire import IFBushfire
+
+THRESH = 331
+
+class TestIFBushfire(unittest.TestCase):
+    """Impact function test"""
+
+    def test_default_values_pass(self, threshold = THRESH):
```
[the remark this answers is not in the document]
Comment (samluethi) [reply climada/entity/impact_funcs/test/test_bf.py:26]: fully re-written the impact_funcs test for wildfire (forgot to update that)

File: climada/entity/impact_funcs/wildfire.py (hunk 3 above)
Comment (samluethi) [reply climada/entity/impact_funcs/wildfire.py:33]: I've made quite many adjustments to the whole impact_funcs/wildfire.
- docstrings
- documentation & harmonisation of impact functions

File: climada/entity/impact_funcs/wildfire.py (hunk 4 above)
Comment (samluethi) [reply climada/entity/impact_funcs/wildfire.py:42]: I've made quite many adjustments to the whole impact_funcs/wildfire.

- docstrings
- documentation & harmonisation of impact functions

File: climada/entity/impact_funcs/wildfire.py (hunk 5 above)
Comment (samluethi) [reply climada/entity/impact_funcs/wildfire.py:51]: I've made quite many adjustments to the whole impact_funcs/wildfire.

- docstrings
- documentation & harmonisation of impact functions

File: climada/entity/impact_funcs/wildfire.py (hunk 6 above)
Comment (samluethi) [reply climada/entity/impact_funcs/wildfire.py:32]: I've made quite many adjustments to the whole impact_funcs/wildfire.

- docstrings
- documentation & harmonisation of impact functions

File: climada/hazard/wildfire.py (hunk 12 above)
Comment (samluethi) [reply climada/hazard/wildfire.py:463]: adjusted the docstring

File: climada/hazard/wildfire.py (hunk 13 above)
Comment (samluethi) [reply climada/hazard/wildfire.py:500]: within set_hist_fire_seasons_FIRMS the function _select_fire_season is called. The methodology there is consistent

File: climada/hazard/wildfire.py (hunk 14 above)
Comment (samluethi) [reply climada/hazard/wildfire.py:597]: done

File: climada/hazard/wildfire.py (hunk 15 above)
Comment (samluethi) [reply climada/hazard/wildfire.py:755]: moved to Centroids

File: doc/tutorial/climada_hazard_Bushfire-Copy1.ipynb (hunk 16 above)
Comment (samluethi) [reply doc/tutorial/climada_hazard_Bushfire-Copy1.ipynb:1]: removed it. will do some further work on the tutorial anyway

File: climada/entity/impact_funcs/wildfire.py (hunk 17 above)
Comment (samluethi) [reply climada/entity/impact_funcs/wildfire.py:30]: fully re-written the impact_funcs -> took id as an input as in tc_impact_funcs

File: climada/entity/impact_funcs/wildfire.py (hunk 18 above)
Comment (samluethi) [reply climada/entity/impact_funcs/wildfire.py]: adjusted

File: climada/entity/impact_funcs/wildfire.py (hunk 19 above)
Comment (samluethi) [reply climada/entity/impact_funcs/wildfire.py]: adjusted

File: climada/entity/impact_funcs/wildfire.py (hunk 20 above)
Comment (samluethi) [reply climada/entity/impact_funcs/wildfire.py]: adjusted

**Pull request merged**